dandiapi/api/tasks.py

```diff
@@ -56,7 +63,81 @@ def write_yamls(version_id: int) -> None:
         logger.info('%s already exists, deleting it', assets_yaml_path)
         storage.delete(assets_yaml_path)
     logger.info('Saving %s', assets_yaml_path)
-    assets_yaml = YAMLRenderer().render([asset.metadata.metadata for asset in version.assets.all()])
+    assets_yaml = YAMLRenderer().render(
+        [asset.generate_metadata(version) for asset in version.assets.all()]
+    )
     storage.save(assets_yaml_path, ContentFile(assets_yaml))
 
     logger.info('Wrote dandiset.yaml and assets.yaml for version %s', version_id)
+
+
+def format_as_index(indices):
+    """Render a JSON schema path as a series of indices."""
+    if not indices:
+        return ''
+    return '[%s]' % ']['.join(repr(index) for index in indices)
+
+
+def format_validation_error(error: jsonschema.exceptions.ValidationError):
+    """
+    Succinctly format a ValidationError.
+
+    The default __str__ for ValidationError includes the entire schema, so we simplify it.
+    """
+    return f'{error.message}\nSee: metadata{format_as_index(error.relative_path)}'  # noqa: B306
+
+
+# TODO what is a reasonable retry delay
+@shared_task(bind=True, default_retry_delay=10)
+@atomic
+# This method takes both a version_id and an asset_id because asset metadata renders differently
+# depending on which version the asset belongs to.
+def validate_asset_metadata(self, version_id: int, asset_id: int) -> None:
+    logger.info('Validating asset metadata for asset %s, version %s', asset_id, version_id)
+    asset = Asset.objects.get(id=asset_id)
+    version = Version.objects.get(id=version_id)
+
+    if asset.status == AssetStatus.VALIDATING.name:
+        # Another task is currently validating, let it finish
+        logger.info('Asset %s is already being validated, wait for it to complete', asset_id)
+        raise self.retry()
+
+    # Begin the validation process so no other validation tasks will run simultaneously
+    asset.status = AssetStatus.VALIDATING.name
+    asset.save()
+
+    try:
+        metadata = asset.generate_metadata(version)
+        if 'schemaVersion' not in metadata:
+            logger.info('schemaVersion not specified in metadata for asset %s', asset_id)
+            asset.status = AssetStatus.INVALID.name
+            asset.validation_error = 'schemaVersion not specified'
+            asset.save()
+            return
+        schema_version = metadata['schemaVersion']
+        schema_url = (
+            'https://raw.githubusercontent.com/dandi/schema/master/'
+            f'releases/{schema_version}/asset.json'
```

Comment:
just a heads up there will be a published-asset.json here, which will also include any fields required for publishing, which are likely to be computed fields anyway.